### PR TITLE
Image syncer

### PR DESCRIPTION
**Pull request by adamwalach**
**Description**

Changes proposed in this pull request:

- ...
- ...
- ...

**Related issue(s)**
https://github.com/kyma-project/test-infra/issues/2426

### CODE REVIEW

**Comment by adamwalach** [comment]
Output:
```
INFO[0000] Source image: quay.io/prometheus/prometheus:v2.17.1
INFO[0002] Source ID: sha256:358a0d2395fe711bb8258e8fb4b2d7865c0a9a6463969bcd1452ee8869ea6653
INFO[0002] Source repo digest: sha256:0d05276441fc3ec287025db43190873fd0088d1b16a45a0012a59f4cbd6e5be9
INFO[0002] Target image: eu.gcr.io/sap-kyma-prow/external/quay.io/prometheus/prometheus:v2.17.1
INFO[0003] Target ID: sha256:358a0d2395fe711bb8258e8fb4b2d7865c0a9a6463969bcd1452ee8869ea6653
INFO[0003] Target repo digest: sha256:0ec13a50dfa65ec1cf8583ef640271e14392534f1d2c21f758eb0e0e441f7540
INFO[0003] Source and target IDs are the same, nothing to do
INFO[0003] ------------------------
INFO[0003] Source image: grafana/grafana:7.0.1
INFO[0006] Source ID: sha256:465c4af545b156014b9a7ca526d5e3577f1b0d5c4a37344ea1d25b0bed89bce4
INFO[0006] Source repo digest: sha256:eb5abc53bed20198052d4ac0e01a8fd3bfe0c80d8e9eb9d021fbc10f93a25514
INFO[0006] Target image: eu.gcr.io/sap-kyma-prow/external/grafana/grafana:7.0.1
INFO[0007] Target image does not exist
INFO[0007] Dry-run mode - tagging and pushing skipped
INFO[0007] ------------------------
INFO[0007] Source image: grafana/lok-error-i:v1.3.0
FATA[0009] image pull failed: Error response from daemon: pull access denied for grafana/lok-error-i, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
exit status 1

**Comment by adamwalach** [comment]
We thought about possibility of using https://github.com/kubernetes-sigs/k8s-container-image-promoter instead of creating our own solution.
Unfortunately "image-promoter" is limited only to Google Container Registry (GCR) - it uses server-side operations that are not supported by other registries like quay.io od Docker Hub.

File: development/image-syncer/main.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+		if dryRun {
+			log.Info("Dry-run mode - tagging and pushing skipped")
+			return nil
+		}
+		if err = cli.ImageTag(ctx, source, target); err != nil {
+			return err
+		}
+		log.Info("Image re-tagged")
+		log.Info("Pushing to target repo")
+		if reader, err := cli.ImagePush(ctx, target, types.ImagePushOptions{RegistryAuth: authString}); err != nil {
```
Comment (Ressetkk) [inline development/image-syncer/main.go:111]: For logging in and tracking errors if the login credentials are wrong can we use for example [RegistryLogin method](https://godoc.org/github.com/docker/docker/client#Client.RegistryLogin) of Docker Client? If the credentials are wrong it returns `unauthorizedError` when the authorisation fails.

Comment (adamwalach) [reply development/image-syncer/main.go:111]: good point, I will test this method against malformed json file

File: development/image-syncer/main.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+		Password: password,
+	}
+	encodedJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return "", err
+	}
+	return base64.URLEncoding.EncodeToString(encodedJSON), nil
+}
+
+func getImageIdAndRepoDigest(ctx context.Context, cli *client.Client, authString, image string) (string, string, error) {
```
Comment (Ressetkk) [inline development/image-syncer/main.go:56]: The `authString` does not seem to be used in the function. Will it have a purpose in a future?

Comment (adamwalach) [reply development/image-syncer/main.go:56]: I tested registries where it was used, but since we don't have a situation like that I will remove that

Comment (adamwalach) [reply development/image-syncer/main.go]: fixed

**Pull request merged**